### ISSUE TITLE
Speedup for usecases of >90% identity hits only

### ISSUE THREAD

**Issue by oschwengers** [bbuchfink/diamond#419]
Hi Oliver, Diamond was not designed for this use case of >90% identity hits only, so I'm pretty sure that substantial speedups would be possible there. Simply building a faster mode that uses longer seeds would be pretty easy, I can look into that in the next days. When it comes to further improving performance for small query numbers, that will take some more time, as I explained.

_Originally posted by @bbuchfink in https://github.com/bbuchfink/diamond/issues/413#issuecomment-740937019_

**Comment by oschwengers** [comment]
@bbuchfink Again, thanks a lot for taking care of this. To keep this request separate from the original, I opened a distinct issue.

FYI: In bacterial genome annotation, *e.g.* [Prokka](https://github.com/tseemann/prokka) & [Bakta](https://github.com/oschwengers/bakta), we often have rather small protein query files (~5k entries for a standard bacterial genome) and run them against larger DBs. Here, sensitivity is not that much of an issue (at least for me) as we often try to identify very distinct protein sequences, sometimes even down to 99% seq identity and high seq coverage (>80%).

Two typical use cases for me:
- diamond blastp vs. UniRef 90: query: ~5k aa seqs, db: ~80,000,000 entries, `--query-cover 80` `--subject-cover 80` `--id 90`
- diamond blastp vs. AMR or IS: query: ~5k aa seqs, db: ~5,000 entries, `--query-cover 90` `--subject-cover 90` `--id 98`

So further optimizations towards these use cases would be very useful for our community and very much appreciated!

**Comment by bbuchfink** [comment]
I have made a first attempt at a faster mode for >90% id hits, using `--fast`. It currently only lives in my dev branch, and you need to set `cmake -DEXTRA=ON` to be able to use it. It will need some further refinement and optimization before official release, but you can give it a try. In your case, it is also advisable to use a large block size (like `-b12` or higher) as well as `-c1`.

Another thing that would help me, what are your typical reporting requirements, do you need a single best hit for each query, a small number of best hits, a full list of hits, etc.?

**Comment by oschwengers** [comment]
Thanks, I'll conduct some performance tests and let you know. 

Regarding your questions:
- On larger servers I already use larger block & chunk sizes. As I'm also developing for tools running on consumer hardware resources are rather tight ( like ~8 Gb netto for Diamond?). 
- I only need the single best hit `--max-target-seqs 1`.

**Comment by bbuchfink** [comment]
There is another experimental option that might help speed things up when using a large database. While Diamond normally computes Smith Waterman extensions on each database block independently, using `--ext-targets N` the best N seed hits will be accumulated over the whole db prior to extension, which can be a lot more efficient when only requiring a small number of best hits. N=100 may be sufficient in your case.

8 GB is certainly possible if you reduce the block size a bit.

**Comment by oschwengers** [comment]
Thanks for taking this on. I tried to compile diamond with `cmake -DEXTRA=ON` on 729ed0aa0512b076c795c0b992cbeade52c5761f  in order to try the new `--fast` option but got the following compile error:
```
diamond/src/stats/matrix_adjust_eigen.cpp: In function ‘void CalculateResiduals(double*, MatrixN&, Vector2Nx&, const double (&)[2], const Vector2NN&, const VectorN&, const VectorN&, const MatrixN&, const Vector2Nx&, double)’:
/var/scratch/oschweng/diamond/diamond/src/stats/matrix_adjust_eigen.cpp:133:53: error: cannot bind non-const lvalue reference of type ‘Block2N&’ {aka ‘Eigen::VectorBlock<Eigen::Matrix<double, 40, 1>, 39>&’} to an rvalue of type ‘Eigen::DenseBase<Eigen::Matrix<double, 40, 1> >::FixedSegmentReturnType<39>::Type’ {aka ‘Eigen::VectorBlock<Eigen::Matrix<double, 40, 1>, 39>’}
  133 |  ResidualsLinearConstraints(resids_z.head<2 * N - 1>(), x, row_sums, col_sums);
```

using the following procedure:
```
git clone https://github.com/bbuchfink/diamond.git
cd diamond
git checkout dev
cmake -DEXTRA=ON
make -j8
```

Is there anything I could try meanwhile?

**Comment by bbuchfink** [comment]
Sorry, please try again with the latest commit.

**Comment by oschwengers** [comment]
Thanks, its working! I performed some tests on a protein query test case (5,456 *ecoli* CDS AA seqs predicted by Prodigal) against a production db:
```
$ diamond dbinfo --db ../db.dmnd 
...
Database format version = 3
Diamond build = 140
Sequences = 79864221
Letters = 24861696439
```

measured via `/usr/bin/time` I tested 4 different setups:
1) Diamond 2.0.6
```
Command being timed: "diamond blastp --threads 32 --db db.dmnd --out test --outfmt 6 --query prots.faa --max-target-seqs 1 --id 90 --query-cover 80 --subject-cover 80"
        User time (seconds): 5935.57
        System time (seconds): 743.17
        Percent of CPU this job got: 1927%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 5:46.48
```

2) Diamond current dev using `--fast`
```
Command being timed: "./diamond/diamond blastp --threads 32 --db db.dmnd --out test --outfmt 6 --query prots.faa --max-target-seqs 1 --id 90 --query-cover 80 --subject-cover 80 --fast"
        User time (seconds): 4978.09
        System time (seconds): 478.64
        Percent of CPU this job got: 1873%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 4:51.32
```

3) Diamond current dev using `--fast` & `--ext-targets`
```
Command being timed: "./diamond/diamond blastp --threads 32 --db db.dmnd --out test --outfmt 6 --query prots.faa --max-target-seqs 1 --id 90 --query-cover 80 --subject-cover 80 --fast --ext-targets 100"
        User time (seconds): 4408.62
        System time (seconds): 586.90
        Percent of CPU this job got: 1713%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 4:51.60
```

4) Diamond current dev using `--ext-targets`
```
Command being timed: "./diamond/diamond blastp --threads 32 --db db.dmnd --out test --outfmt 6 --query prots.faa --max-target-seqs 1 --id 90 --query-cover 80 --subject-cover 80 --ext-targets 100"
        User time (seconds): 3558.81
        System time (seconds): 657.95
        Percent of CPU this job got: 1624%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 4:19.52
```

So the new `--fast` mode results in a reasonable runtime reduction but the `--ext-targets` flag alone reduces the overall runtime even more.  Using `--fast`, the the `--ext-targets` has no further impact.
So both provide very good results in terms of runtime reductions. I'm eager to see diamond further following this path. Please, let me know If I can test further versions and thanks again!

**Comment by bbuchfink** [comment]
Glad that it helped, I have been working on some more improvements for small input files and hopefully will be able to release this in the near future. You should also be able to increase the performance by using `-c1` and changing the block size (`-b`) according to your RAM.

**Comment by oschwengers** [comment]
Thanks for the brand new release (v2.0.8). Could you briefly comment on whether there is already some usable progress regarding the speed optimization for "high-identity" queries?

**Comment by bbuchfink** [comment]
No, not at the moment, but I'm working on it now. Progress should happen over the next weeks.

**Comment by oschwengers** [comment]
Cool, thanks a lot! I'm looking forward to this :-)

**Comment by bbuchfink** [comment]
I have now implemented various optimizations to improve performance for small query files. It is still preliminary work, and only available in my `dev` branch at the moment. It is designed to work together with BLAST databases and take advantage of memory mapping (`dmnd` database files do not support mmap at this time).

BLAST databases have to be prepared using this call first: `diamond prepdb -d dbpath`

Searching one E.coli genome runs in 4.6 minutes against the full NR (370 million seqs) on a 24-core machine using this call:

`diamond blastp -q GCF_000008865.2_ASM886v2_protein.faa.gz -d /mnt/hd1/bbuchfink/ncbi/nr/nr -o out -c1 -t /dev/shm --ext-targets 300 --fast`

Only works if the host has enough RAM to cache the whole database in memory. May need several warmup runs until the database is fully cached.

**Comment by oschwengers** [comment]
Thanks a lot for these promising developments!

Is the support of `mmap` for regular `dmnd` DBs on your ToDo?
I've developed a couple of tools which exclusively use Diamond and thus have several bundled `dmnd` DBs. Hence, I'm rather reluctant to create Blast DBs for production. However, of course, I'll test this to get an idea on potential speedups. Very much appreciated!

**Comment by oschwengers** [comment]
I get a compilation error:

```
git clone https://github.com/bbuchfink/diamond.git
cd diamond
git checkout dev
cmake -DEXTRA=ON
make -j8
...
diamond/src/run/main.cpp:190:4: error: ‘prep_blast_db’ was not declared in this scope
  190 |    prep_blast_db();
```

**Comment by bbuchfink** [comment]
Sorry, it only compiles with BLAST db support (see compilation instructions). The `dev` branch is a bit messy at the moment and does not work with the regular dmnd files.

**Comment by bbuchfink** [comment]
> Is the support of `mmap` for regular `dmnd` DBs on your ToDo?

I will probably implement this too, but it will be less efficient since dmnd files store the sequence ids together with the sequences, so both have to be cached it memory.

**Comment by bbuchfink** [comment]
The improvements are now available in the v2.0.11 release. I also added `--algo ctg` to enable pre-v2.0.9 behaviour. See here for a benchmark:

https://github.com/bbuchfink/diamond/wiki/Small-query-benchmark

This is with respect to the NR (400 million seqs), the runtime should be linear in the size of the database.

**Comment by oschwengers** [comment]
I've done a few runtime benchmarks and I can confirm that for our usecases, `2.0.11` is indeed remarkably faster (like ~10% faster than `2.0.10` and 100% faster than `2.0.9`).

Thanks a lot for taking care of this! Great job and indeed very much appreciated!
I'll close this.
